### PR TITLE
apply spatial index lookup

### PR DESCRIPTION
**Pull request by JCOR11599**
Changes:
 - Add support for displaing `SPATIAL KEY` for `SHOW CREATE TABLE`.
 - Allow `IndexedTableAccess` for in-memory tables that have spatial indexes
 - dolt pr with interface changes: https://github.com/dolthub/dolt/pull/5164
 - added new testing suite for Spatial Indexes
 - some fixes to prepared queries for filtered tables

### CODE REVIEW

File: memory/table.go [hunk 1]
```diff
@@ -450,6 +496,78 @@ func (i *tableIter) getFromIndex(ctx *sql.Context) (sql.Row, error) {
...
+		(gMinX <= i.maxX && i.maxX <= gMaxX) ||
+		(i.minX <= gMinX && gMinX <= i.maxX) ||
+		(i.minX <= gMaxX && gMaxX <= i.maxX)
+	yInt := (gMinY <= i.minY && i.minY <= gMaxY) ||
+		(gMinY <= i.maxY && i.maxY <= gMaxY) ||
+		(i.minY <= gMinY && gMinY <= i.maxY) ||
+		(i.minY <= gMaxY && gMaxY <= i.maxY)
+	if !(xInt && yInt) {
+		return i.Next(ctx)
+	}
```
Comment (max-hoffman) [inline memory/table.go:528]: maybe logic that should be generalized for implementor use?

File: enginetest/queries/script_queries.go [hunk 2]
```diff
@@ -3191,104 +3194,375 @@ var SpatialScriptTests = []ScriptTest{
...
+				Expected: []sql.Row{
+					{types.Point{}},
+				},
 			},
 			{
-				Query:       "create table geom(g multipoint not null srid 4326, spatial index (g))",
-				ExpectedErr: sql.ErrUnsupportedSpatialIdx,
+				Query: "explain select p from point_tbl where st_intersects(p, point(0,0))",
+				Expected: []sql.Row{
+					{"Filter"},
```
Comment (max-hoffman) [inline enginetest/queries/script_queries.go:3290]: plan tests need to go in their own codegen-able section

File: sql/analyzer/indexes.go [hunk 3]
```diff
@@ -261,6 +262,59 @@ func getIndexes(
 		return result, nil
 	}
 
+	// TODO: maybe this block can just go into previous switch
+	// TODO (james): add all other spatial index supported functions here
+	// TODO: make generalizable to all functions?
+	switch e := e.(type) {
+	case *spatial.Intersects:
```
Comment (max-hoffman) [inline sql/analyzer/indexes.go:269]: missing `getNegatedIndexes` and `getMultiColumnIndexes`, or if not supporting at least test those to make sure they don't break

File: memory/table.go [hunk 4]
```diff
@@ -450,6 +496,78 @@ func (i *tableIter) getFromIndex(ctx *sql.Context) (sql.Row, error) {
...
+
+	if len(i.columns) > 0 {
```
Comment (max-hoffman) [inline memory/table.go:530]: could maybe put this check at the beginning, avoid all that unnecessary work

File: memory/table.go [hunk 5]
```diff
@@ -1002,6 +1120,29 @@ func (t *IndexedTable) LookupPartitions(ctx *sql.Context, lookup sql.IndexLookup
...
+		minPoint, ok := lower.(types.Point)
+		if !ok {
+			return nil, sql.ErrInvalidGISData.New()
+		}
+		maxPoint, ok := upper.(types.Point)
+		if !ok {
+			return nil, sql.ErrInvalidGISData.New()
+		}
+
+		ord := lookup.Index.(*Index).Exprs[0].(*expression.GetField).Index()
```
Comment (max-hoffman) [inline memory/table.go:1135]: do spatial indexes support nary columns?

File: sql/index_builder.go [hunk 6]
```diff
@@ -285,9 +285,19 @@ func (b *IndexBuilder) updateCol(ctx *Context, colExpr string, potentialRanges .
 		return
 	}
 
+	if b.idx.IsSpatial() {
+		if _, ok := currentRanges[0].LowerBound.(BelowNull); ok {
```
Comment (max-hoffman) [inline sql/index_builder.go:289]: This method is supposed to intersect two arbitrary nary sets of current and new ranges. This addition appears to intersect single dimension ranges for one specific edge case: 1) when there is only one current and potential range expression, 2) and we can take the current range lower and new range upper bound. I don't think this is correct, can we make it do a general intersection, or not use lookup builder for spatial indexes if this logic is correct and they are subject to very different range intersection rules?

File: sql/index_builder.go [hunk 7]
```diff
@@ -243,7 +243,7 @@ func (b *IndexBuilder) Ranges(ctx *Context) RangeCollection {
 			b.err = err
 			return nil
 		}
-		if !isempty {
+		if !isempty || b.idx.IsSpatial() {
```
Comment (max-hoffman) [inline sql/index_builder.go:246]: does spatial indexes IsEmpty() not return something sensible? i'd maybe prefer the interface to be a reliable indicator of empty

File: sql/index.go [hunk 8]
```diff
@@ -93,8 +95,9 @@ type IndexLookup struct {
 	// values; the range is null safe, the index is unique, every index
 	// column has a range expression, and every range expression is an
 	// exact equality.
-	IsPointLookup bool
-	IsEmptyRange  bool
+	IsPointLookup   bool
+	IsEmptyRange    bool
+	IsSpatialLookup bool
```
Comment (max-hoffman) [inline sql/index.go:100]: are there any details about spatial index lookups that are not obvious any maybe helpful to document?

File: sql/analyzer/pushdown.go [hunk 9]
```diff
@@ -460,6 +460,9 @@ func getPredicateExprsHandledByLookup(ctx *sql.Context, a *Analyzer, idxTable *p
 	if !ok {
 		return nil, nil
 	}
+	if filteredIdx.IsSpatial() {
```
Comment (max-hoffman) [inline sql/analyzer/pushdown.go:463]: are spatial lookups lossy? what scenarios do we need to keep filter vs not?

**Review by max-hoffman — CHANGES_REQUESTED** [review]
I think this did a good job of getting a swath of spatial index functionality working. Some things I would like to see before merging:

1) Tests for the subset of IndexedTableAccess behaviors that spatial indexes do not support but need to fail in sensible ways. Lookup joins, concat joins, multi column expression lookups, intersecting multiple filters, negated filters.

2) Tidier edge case handling for spatial edge cases. There are a bunch of hacks that look like they make the current tests work, but that might fail with testing in (1).

3) Documentation for explaining special cases where spatial indexes do need to be treated differently, as opposed to the carve-outs for incremental progress (which could also maybe use comments).

File: sql/analyzer/pushdown.go (hunk 9 above)
Comment (JCOR11599) [reply sql/analyzer/pushdown.go:463]: Spatial lookups are lossy, so we definitely should keep the filter.
They're basically doing a cheap bounding box check, before running in actual expensive st_intersect/st_equals/etc function.

File: memory/table.go (hunk 5 above)
Comment (JCOR11599) [reply memory/table.go:1135]: Spatial indexes can only be over exactly one geometry column
[MySQL docs](https://dev.mysql.com/doc/refman/8.0/en/create-index.html#create-index-spatial:~:text=Available%20for%20single%20spatial%20columns%20only.%20A%20spatial%20index%20cannot%20be%20created%20over%20multiple%20spatial%20columns.)

Comment (max-hoffman) [reply memory/table.go:1135]: in that case, would it make more sense to have a separate SpatiaIndexBuilder?

File: sql/analyzer/pushdown.go (hunk 9 above)
Comment (max-hoffman) [reply sql/analyzer/pushdown.go:463]: it would be good to make this info and its life cycle peculiarities more obvious

File: sql/index_builder.go (hunk 7 above)
Comment (JCOR11599) [reply sql/index_builder.go:246]: `Range.IsEmpty()` will call `RangeColumnExpr.IsEmpty()`, which checks if `UpperBound >= LowerBound`.
It's possible for the bounding box to be just a point (`UpperBound == LowerBound`), which is incorrectly classified as empty.

A better fix might be to just have a `SpatialIndexBuilder` with different logic like you mentioned before.

File: sql/analyzer/pushdown.go (hunk 9 above)
Comment (JCOR11599) [reply sql/analyzer/pushdown.go:467]: does this mean add a comment?

File: enginetest/spatial_index_tests.go [hunk 10]
```diff
@@ -0,0 +1,453 @@
...
+
+			"insert into geom_tbl values (point(2,-2))",
+			"insert into geom_tbl values (point(2,-1))",
+			"insert into geom_tbl values (point(2,0))",
+			"insert into geom_tbl values (point(2,1))",
+			"insert into geom_tbl values (point(2,2))",
+		},
+		tests: []SpatialIndexPlanTestAssertion{
+			{
+				skipPrep: true,
```
Comment (max-hoffman) [inline enginetest/spatial_index_tests.go:134]: why do so many of these need prepareds skipped?

File: enginetest/spatial_index_tests.go [hunk 11]
```diff
@@ -0,0 +1,453 @@
...
+	exp      []sql.Row
+}
+
+type SpatialIndexPlanTest struct {
+	name  string
+	setup []string
+	tests []SpatialIndexPlanTestAssertion
+}
+
+var SpatialIndexTests = []SpatialIndexPlanTest{
```
Comment (max-hoffman) [inline enginetest/spatial_index_tests.go:44]: i don't see any tests with compound filters -- AND, OR, etc. How do we choose between two choices for filtered index access? For two spatial options, but also a spatial/regular index choice. Can we intersect (AND) two spatials/spatial+regular? Similar treatment w/ concat nodes and OR? Maybe a test with filters that should be pushed into subqueries too, like `select st_aswkt(g) from (select * from geom_tbl) t where st_intersects(g, point(0,0))`

File: sql/analyzer/indexes.go [hunk 12]
```diff
@@ -240,6 +241,10 @@ func getIndexes(
 		result := multiColumnIndexes
 		// Next try to match the remaining expressions individually
 		for _, e := range unusedExprs {
+			// TODO: eventually support merging spatial lookups
+			if _, ok := e.(*spatial.Intersects); ok {
```
Comment (max-hoffman) [inline sql/analyzer/indexes.go:245]: could maybe use an interface for checking if an expression is a spatial function

**Review by max-hoffman — APPROVED** [review]
LGTM, thanks for adding the additional tests and prepared statement fixes
